benchflow-test-manager/application/src/main/java/cloud/benchflow/testmanager/scheduler/TestTaskScheduler.java

```diff
@@ -221,281 +245,115 @@ private synchronized void handleTestRunningState(String testID) {
-
-    RemoveNonReachableExperimentsTask task = new RemoveNonReachableExperimentsTask(testID);
+  private void handleTerminatedState(String testID) {
+    // remove max running time timeout
+    ScheduledFuture timeoutTaskFuture = timeoutTasks.remove(testID);
+    timeoutTaskFuture.cancel(true);
 
-    // TODO - should go into a stateless queue (so that we can recover)
-    Future<?> future = taskExecutorService.submit(task);
+    // remove any tasks left
+    testTasks.remove(testID);
 
-    // replace with new task
-    testTasks.put(testID, future);
-
-    waitForRunningTaskToComplete(testID, future, DETERMINE_EXECUTE_EXPERIMENTS);
+    // remove test from running queue so dispatcher can run next test
+    runningQueue.remove(testID);
   }
 
-  private void validateTerminationCriteria(String testID) {
-
-    logger.info("validateTerminationCriteria with testID: " + testID);
 
-    ValidateTerminationCriteria task = new ValidateTerminationCriteria(testID);
+  /**
+   * Called when user terminates test or max run time has been reached.
+   */
+  public synchronized void terminateTest(String testID) {
 
-    // TODO - should go into a stateless queue (so that we can recover)
-    Future<TerminationCriteriaResult> future = taskExecutorService.submit(task);
-
-    // replace with new task
-    testTasks.put(testID, future);
+    logger.info("terminateTest: " + testID);
 
     try {
 
-      TerminationCriteriaResult result = future.get();
-
-      switch (result) {
+      BenchFlowTestState testState = testModelDAO.getTestState(testID);
 
-        case GOAL_NOT_REACHABLE:
-          testModelDAO.setTestState(testID, TERMINATED);
-          testModelDAO.setTestTerminatedState(testID, COMPLETED_WITH_FAILURE);
+      switch (testState) {
+        case READY:
+          // remove from ready queue
+          readyQueue.remove(testID);
+          // set to terminated
+          testModelDAO.setTestState(testID, BenchFlowTestState.TERMINATED);
+          testModelDAO.setTestTerminatedState(testID, TestTerminatedState.PARTIALLY_COMPLETE);
+          handleTestState(testID);
           break;
 
-        case GOAL_REACHABLE_REGRESSION_PREDICTION_ACCEPTABLE:
-        case GOAL_REACHABLE_NO_REGRESSION_EXPERIMENTS_EXECUTED:
-          testModelDAO.setTestState(testID, TERMINATED);
-          testModelDAO.setTestTerminatedState(testID, GOAL_REACHED);
+        case WAITING:
+        case START:
+          // set to terminated
+          testModelDAO.setTestState(testID, BenchFlowTestState.TERMINATED);
+          testModelDAO.setTestTerminatedState(testID, TestTerminatedState.PARTIALLY_COMPLETE);
+          handleTestState(testID);
           break;
 
-        case GOAL_REACHABLE_REGRESSION_PREDICTION_NOT_ACCEPTABLE:
-          testModelDAO.setTestRunningState(testID, REMOVE_NON_REACHABLE_EXPERIMENTS);
+        case RUNNING:
+
+          TestRunningState runningState = testModelDAO.getTestRunningState(testID);
+
+          // set test to terminated
```

Comment:
Worked on this in https://github.com/benchflow/benchflow/pull/474#commits-pushed-08d26a0